Adds in_toto_record {start, stop} command + minor changes/fixes

lukpueh opened the pull request:
In some cases it is not feasible to wrap a single command with toto-run, e.g. when allowedly editing a bunch of files shortly before building.
To not disrupt the in-toto supply chain we provide two additional cli tools that, where one only records materials to a temporary file and the second records products and finalizes the link metadata file.

- Adds new command line tool (with two sub commands) to perform a multi-part in-toto step
- Adds unittests for in_toto_record start/stop (requires `mock` package)
- Adds default values to optional parameters of runlib.create_link_metadata
- Changes `attr` class attribute defaults to use `attr.Factory` in order to circumvent strange memory re-use (see. https://github.com/in-toto/in-toto/commit/be7619173e13cbfb17093ac945a16083eb3d12db commit message for more details).
- Fixes wrong default type for `layout.keys` attributes (https://github.com/in-toto/in-toto/issues/41)
- Removes two obsolete utility classes:  
 - `GenericPathList` 
was used to *normalize* relative artifact paths at verification, which now happens at recording 
 - `ComparableHashDict`
was used to compare artifact hashes with multiple hash algorithms, deprecated by `util.flatten_and_invert_artifact_dict`

lukpueh committed:
Adds defaults runlib.create_link_metadata args

It makes sense to create relatively empty Link metadata files
e.g. when splitting toto-run by first only recording materials,
then running the command manually and finally recording products
to a Link metadata file

lukpueh committed:
Adds toto_start_record and toto_stop_record (WIP)

In some cases it is not feasible to wrap a single command with
toto-run, e.g. when allowedly editing a bunch of files shortly before
building.
To not disrupt the in-toto supply chain we provide two additional cli
tools that, where one only records materials to a temporary file
and the second records products and finalizes the link metadata
file.

lukpueh committed:
Adds factories, removes util classes

- Adds default factories for lists and dictionaries to in-toto model
attributes. This fixes a strange python/attrs behavior that
re-uses previously allocated memory space for e.g. lists when initializing
them with an empty default list:
```
import attr
@attr.s(repr=False)
class Foo(object):
  bar = attr.ib([])

Foo().bar.append("foobar")
Foo().bar  == ["foobar"]
```

- Removes obsolete util classes
  GenericPathList
  ComparableHashDict

- Fixes layout.keys default type
  c.f. https://github.com/in-toto/in-toto/issues/41

lukpueh committed:
Adds in_toto_record as separate cli tool + tests

- Moves in_toto_record_start and in_toto_record_stop from
toto-run to separate cli tool with two subcommands
- Basic usage is:
  ```
  in_toto_record.py start --step-name edit-files --materials . --key bob
  # Edit files manually ...
  in_toto_record.py stop --step-name edit-files --products . --key bob
  ```
- Adds unittests

lukpueh committed:
Fixes document docstring in_toto_record

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+        "test-key"
+    ]
+    self.test_artifact = "test-artifact"
+    open(self.test_artifact, "a").close()
```
what about doing this in a tmpdir to avoid polluting $PWD?

lukpueh replied:
Good idea! Will do.

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+    files_to_remove = [
+      "test-key", "test-key.pub", "test-step.link", self.test_artifact
+    ]
+    _try_remove_files(files_to_remove)
```
same here, we could replace this with `shutil.rmdir`

lukpueh replied:
Ditto

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+
+
+
+
```
This is too much whitespace. Could you remove it please?

lukpueh replied:
Will do

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+
+    with patch.object(sys, 'argv', self.args + ["stop", "--products",
+        self.test_artifact]):
+      in_toto_record_main()
```
I wonder if we need this use of mock, or if it's just a byproduct of a weird architecture.  Wouldn't it be easier if we had a function to call? e.g., `start_recording(**kwargs)` and `stop_recording(**kwargs)`

SantiagoTorres replied:
Now I see, maybe we could have "main()" take the sys.* arguments instead?

lukpueh replied:
What do you mean by "main() taking the sys.* arguments"? Something like`main(sys.argv)`? I'm not sure if the architecture is weird. Let me know if you have architectural suggestions.

SantiagoTorres replied:
well, you need specific info from argv right? (e.g., a list of products, a list of materials, etc.)

lukpueh replied:
The reason why I use `mock` here is that I want to test the argparser part of the cli tool.

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+    generate_and_write_rsa_keypair(self.key_path)
+
+    self.test_product = "test-product"
+    open(self.test_product, "a").close()
```
why do we open it in "append" mode?

lukpueh replied:
No reason. Will change to `w`

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+    files_to_remove = [
+      self.key_path, self.key_path + ".pub", self.key_path2,
+      self.key_path2 + ".pub", self.test_product, self.step_name + ".link",
+      "." + self.step_name + ".link-unfinished"
```
I think it'd be better if you used `os.path.join()` here

lukpueh replied:
Why? This is only a list of file names. Anyway, I will change this to use a tmp directory as you suggested earlier.

SantiagoTorres replied:
ah right, I also think format makes it more readable

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+  <Side Effects>
+    Loads private key from disk
+    Creates a file with  name ".<step_name>.link-unfinished"
+    Calls sys.exit(1) if an exception is raised
```
would it be better to return with 127?

lukpueh replied:
"command not found"? Does not seem suitable to me.

SantiagoTorres replied:
d'oh. You're right. I was reading the [glibc docs](https://www.gnu.org/software/libc/manual/html_node/Exit-Status.html) and it irks me that 1 can mean "everything ok but unwanted output" (e.g., in the case of diff). So maybe 2 is unambiguous?

This is a minor nit though. I guess 1 is also a good choice.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+
+  """
+
+  unfinished_fn = "." + str(step_name) + ".link-unfinished"
```
can we use format for this?

lukpueh replied:
Yes we can. Will change.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+    log.doing("store metadata to '{0}'...".format(unfinished_fn))
+    link.dump(unfinished_fn)
+
+    return link
```
Why  is the lnk returned after dumping?

lukpueh replied:
TBH, I did this to inspect the link object in a unittest. But know that you mention it it does not seem so smart anymore.

SantiagoTorres replied:
I mean, it's relatively convenient. I was curious as to why, maybe it's good interface and we can mention it as the return value in the docs?

lukpueh replied:
I will change this because the function should only be called by the command line tool.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+
+  <Exceptions>
+    Exception if signature fails
+    IOException if file ".<step_name>.link-unfinished" can't be read
```
or written, I assume.

lukpueh replied:
`in_toto_record_stop` does not write to the file. It only reads it and then removes it.

lukpueh replied:
I guess we would get an IOException if we don't have write access to the parent directory. Should we mention that?

lukpueh replied:
Actually that would be an `OSError`.

SantiagoTorres replied:
Yep, that's what I meant. OSError (I thought it was IOError) when you can't write the finished link.

lukpueh replied:
I just realize that `in_toto_record_start` and `in_toto_record_stop` always raise a `SystemExit` exception. These are not library functions but should only be called by the command line tool.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+  try:
+    log.doing("load link signing key...")
+    key = toto.util.prompt_import_rsa_key_from_file(key_path)
+    keydict = {key["keyid"] : key}
```
does this make the keyid be repeated inside the value? maybe we don't want to have this.

lukpueh replied:
Why not, this is `ssl_crypto.formats.KEYDICT_SCHEMA`

SantiagoTorres replied:
is it? I find it surprising that we need to key using the keyid and then keep the keyid inside the value also. This isn't something we should discuss in this PR though.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+    link.products = runlib.record_artifacts_as_dict(product_list)
+
+    log.doing("update signature...")
+    link.signatures = []
```
I thought we didn't need this anymore, do we?

lukpueh replied:
This is different. Here we load an unfinished link object and expect it to already have a signature (gets verified above). Then we update the object, clear the existing signatures and re-sign.

SantiagoTorres replied:
Oh, right 👍

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+               .format(lpad))
+
+  in_toto_args = parser.add_argument_group("Toto options")
+  # FIXME: Name has to be unique!!! Where will we check this?
```
I don't think we can check this here.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,221 @@
+
+  in_toto_args = parser.add_argument_group("Toto options")
+  # FIXME: Name has to be unique!!! Where will we check this?
+  # FIXME: Do we limit the allowed characters for the name?
```
Good point, this is not on the specification, so I need to consider it. I wouldn't worry about this in here though.

SantiagoTorres commented on toto/models/common.py:
```diff
@@ -73,7 +71,7 @@ class Signable(Metablock):
   pretty printed JSON string not containing the signatures attribute) and store
   the signature (signature format: ssl_crypto__formats.SIGNATURE_SCHEMA) """
-  signatures = attr.ib([])
+  signatures = attr.ib(default=attr.Factory(list))
```
👍

SantiagoTorres commented on toto/models/layout.py:
```diff
@@ -69,17 +69,18 @@ class Layout(models__common.Signable):
     keys:
-        a list of public keys used to verify the signature of link
-        metadata file related to a step
+        A dictionary of  public keys used to verify the signature of link
```
Minor nit: remove the double space please.

lukpueh replied:
Will do

SantiagoTorres commented on toto/runlib.py:
```diff
@@ -177,8 +177,8 @@ def execute_link(link_cmd_args, record_byproducts):
-def create_link_metadata(link_name, materials_dict, products_dict,
-    link_cmd_args, byproducts, return_value):
+def create_link_metadata(link_name, materials_dict={}, products_dict={},
+    link_cmd_args="", byproducts={}, return_value=None):
```
do we want to have default values here? or default to non and use ifs instead?

lukpueh replied:
Yes we do, because in some cases (e.g. `in_toto_record_start`) it makes sense to create an almost empty link metadata. I thought about your concern (especially given `attrs` fix in this PR) but if you look at the function the default data structures don't get updated. So we should't have a problem here.

SantiagoTorres replied:
Yeah, I guess this is a non-issue. I can't think of an example in which create_link_metadata is called twice in the same process.

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+    self.key_path = "test-key"
+    self.key_path2 = "test-key2"
+    generate_and_write_rsa_keypair(self.key_path)
+    generate_and_write_rsa_keypair(self.key_path)
```
did you mean key_path2 here?

lukpueh replied:
Well spotted, thanks!

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+from toto.in_toto_record import in_toto_record_start, in_toto_record_stop
+
+# Suppress all the user feedback that we print using a base logger
+logging.getLogger().setLevel(logging.CRITICAL)
```
Maybe it would be convenient to have a "disable" call on the logger that replaces everything with a null handler.

lukpueh replied:
I don't know if this changes much regarding convenience. In any case I'd like to rethink the logging/user feedback architecture with https://github.com/in-toto/in-toto/issues/6#issuecomment-259750991.

SantiagoTorres reviewed (changes requested):
It overall looks good. Most of my stuff were minor nits.

lukpueh committed:
Squeeze double whitespace in layout docstring

lukpueh committed:
Use temp dir in in_toto_record tests + misc changes

- Write files created while testing in_toto_record.py to a
  temporary directory.
- Change in_toto_record_start and in_toto_record_stop to return
  None (these are CLI facing functions).
- Fix unittest key bug
- Remove newlines
- Use {}.format instead of string concat
- Replace "a" with "w" file mode for files that get created in
  unittests
- Fix Exceptions docstring for in_toto_record_start and
  in_toto_record_stop

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+
+    with patch.object(sys, 'argv', self.args + ["stop", "--products",
+        self.test_artifact]):
+      in_toto_record_main()
```
Speaking of testing argparse, woudln't we want to test with extraneous input (e.g., wrong flags?)

lukpueh replied:
Okay, I can do this. The *problem* is that argparse does not raise meaningful exceptions per default, it just prints something like e.g. `in_toto_record.py: error: too few arguments` to the console and then exits. I see two ways:
- override arpgarse's `error()` method to raise meaningful exceptions which I can assert
- assert `SystemExit` (plus maybe buffer output in unittest and grep for argparse error message)

Any strong feelings about either one?

SantiagoTorres replied:
Hmm, I'm not sure, to me it feels like testing argparse is not completely necessary if we have a proper architecture. Imho, doing the following would suffice:

1.- change main() to receive explicit arguments
2.- test main to check for proper arguments
3.- Move the argparse function to a function that does that, parse args
4.- Test that function separately

lukpueh replied:
`in_toto_record_start` and `in_toto_record_stop` already receive those explicit arguments and are tested separately. `main` just does the parsing and calls either start or stop.
I think we can leave main like it is. If we want to test the argparser for wrong arguments we still need to decide between the two approaches (check for custom ArgparseException or SystemExit).

SantiagoTorres replied:
So this is what I don't understand, this is a script that has a `main` function and only runs that `main` function. Why do we declare `main` in the first place? is it going to be called from somewhere else (e.g., `import x; x.main()`)? if we do this, then would a caller need to mock sys also? Wouldn't it be more explicit to have main's functionality inside the if? My feeling is that this hinders functional separation, and that's why we are needing these mocks.

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+  def test_create_link_unfinished_file(self):
+    """Test record start creates a link metadata file. """
+    in_toto_record_start(self.step_name, self.key_path, [self.test_material])
+    open(self.link_name_unfinished)
```
do we want to do something a little bit more thorough here?

lukpueh replied:
Do you have something in mind? I want the individual function that are called in `test_create_link_unfinished_file` to get their own thorough unittests.

SantiagoTorres replied:
e.g., something is saved, is that all we need to know? maybe we want to check that it is an actual piece of link metadata (maybe use the read method?). This test passes if you replace `in_toto_record_start()` with `touch $filename`

SantiagoTorres commented on test/test_in_toto_record.py:
```diff
@@ -0,0 +1,185 @@
+  def test_missing_unfinished_file(self):
+    """Test record stop exits on missing unfinished file. """
+    with self.assertRaises(SystemExit):
+      in_toto_record_stop(self.step_name, self.key_path, [])
```
Is this test assuming that the tests run in order? that may not always be the case iirc. Maybe we want to ensure that there is no unfinished file before the `with self.assertRaises()`

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,215 @@
+    log.doing("store metadata to '{0}'...".format(unfinished_fn))
+    link.dump(unfinished_fn)
+
+  except Exception, e:
```
`Exception as e`, please

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,215 @@
+  unfinished_fn = ".{}.link-unfinished".format(step_name)
+
+  try:
+    log.doing("Start recording '{0}'...".format(step_name))
```
Maybe we could use present continuous instead of simple present?
```
"recording {0} ..."
"recording materials..."
"creating preliminary link metadata"
"signing metadata"
"storing metadata in {0}"
```

lukpueh replied:
Sounds good to me. Let's fix this with https://github.com/in-toto/in-toto/issues/6#issuecomment-264205129

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,215 @@
+
+  """
+
+  unfinished_fn = ".{}.link-unfinished".format(step_name)
```
Maybe we could keep this constant string as a module_level global
```python
UNFINISHED_FILENAME_FORMAT = ".{step-name}.link-unfinished"
...
def ...
  unfinished_fn = UNFINISHED_FILENAME_FORMAT.format(step_name=step_name)
```

lukpueh replied:
Good idea.

SantiagoTorres commented on toto/in_toto_record.py:
```diff
@@ -0,0 +1,215 @@
+    log.doing("remove unfinished file...")
+    os.remove(unfinished_fn)
+
+  except Exception, e:
```
exception as e please

lukpueh replied:
Will do

SantiagoTorres reviewed (changes requested):
I think it's pretty much there, thanks!

lukpueh committed:
Restructures in_toto_record, adds verbose flag

- in_toto_record consists of a command line argument parser (main)
  with two subparsers, and a function for each respective subparser,
  which just call into their equivalent runlib function that in turn
  actually implements the workflow
  This architecture is dicussed in more detail in
  https://github.com/in-toto/in-toto/issues/6#issuecomment-263988714

- adds a --verbose flag for in_toto_record .. {start,stop} ...
  which sets the logging level to logging.INFO
  logging.INFO is currently the default but can be raised to e.g.
  logging.WARNING once the other in-toto command line tools
  also get a --verbose flag

SantiagoTorres commented:
LGTM 💯
